### PR TITLE
Replace docker/libtrust with go-jose/go-jose

### PR DESCRIPTION
**Pull request by milosgajdos**
This PR replaces all the [docker/libtrust](https://github.com/docker/libtrus) JWT machinery with [go-jose/go-jose](https://github.com/go-jose/go-jose) module. Some of the code has been adopted from the `libtrust` and adjusted for some of the use cases covered by the token authorization code, especially in the unit tests.

**NOTE:** [docker/libtrust](https://github.com/docker/libtrust) module has been archived for several years now.

Closes https://github.com/distribution/distribution/issues/1820

### CODE REVIEW

**Comment by thaJeztah** [comment]
So from the discussion on slack; did we actually still need it at all? /cc @dmcgowan

**Comment by thaJeztah** [comment]
Oh! Should've looked; it somehow found it's way into auth as well, not just for signing v1 images

**Comment by milosgajdos** [comment]
It is indeed. This PR is in a draft because its a ghetto code that needs cleaning up.

File: docs/spec/auth/jwt.md
Comment (Jamstah) [inline docs/spec/auth/jwt.md:71]: How many auth server implementation actually use this kid format? Can we deprecate the libtrust kid format?

File: registry/auth/token/crypto.go
Comment (Jamstah) [inline registry/auth/token/crypto.go:16]: Does this need exporting? Do any of its methods need exporting?

File: registry/auth/token/token_test.go
Comment (Jamstah) [inline registry/auth/token/token_test.go:2]: I'd be tempted to make this two commits, the first one for the implementation, the second for the tests. That way we know we pass the current tests before changing them to use the new library.

File: registry/auth/token/certificates.go
Comment (milosgajdos) [reply registry/auth/token/certificates.go:1]: I agree, I think we could rename it to `certificates_test.go` so it doesn't get compiled into the binary 🤔

File: registry/auth/token/token_test.go
Comment (milosgajdos) [reply registry/auth/token/token_test.go:2]: Funny you should mention that - that's literally how I went about it 😄 I started with the implementation first, got the tests passing with `libtrust` module, then modified the tests. Happy to do this in 2 steps if folks want me to.

File: go.mod
Comment (milosgajdos) [reply go.mod:17]: `go-jose` is the most comprehensive; for example `golang-jwt` doesn't have "native" support for JWKs support (only via "extenstion" see https://pkg.go.dev/github.com/golang-jwt/jwt/v5#section-readme) for which is mandatory in distribution. Besides the API and implementation are not ideal either. The other library I had a look at was https://github.com/lestrrat-go/jwx which is better than `golang-jwt`, but still less comprehensive with more laborious API that'd require writing way more code than what `go-jose` does (square inc folks designed it pretty well, bar a few things I'd like to change there - maybe future PRs :))

File: registry/auth/token/token_test.go
Comment (Jamstah) [reply registry/auth/token/token_test.go:2]: Might be a good proof point, but I'm happy if you say you did it :)

File: registry/auth/token/certificates.go
Comment (Jamstah) [reply registry/auth/token/certificates.go:1]: I don't think it would get compiled in anyway if its unused, its more about making sure that future developers don't think its an interface that needs to be supported.

Just not exporting anything would be a good start.

File: go.mod
Comment (Jamstah) [reply go.mod:17]: OK. I'm happy with either really, just wanted to flag that it seemed like a small project so could end up being less secure (less eyes) or need replacing (if people migrate to golang-jwt).

File: registry/auth/token/certificates.go
Comment (milosgajdos) [reply registry/auth/token/certificates.go:1]: > I don't think it would get compiled in anyway if its unused

I am intrigued to learn if this is the case given it would have been in a package 🤔 Unfortunately my Go compiler knowledge is lacking here.

Can defo unexport, but I really do think we should consider chucking it into a test file 🤔

Comment (Jamstah) [reply registry/auth/token/certificates.go:1]: https://www.reddit.com/r/golang/comments/6vbh65/are_unused_functions_compiled/?rdt=58347

Removed by the linker apparently.

File: docs/spec/auth/jwt.md
Comment (milosgajdos) [reply docs/spec/auth/jwt.md:71]: To be honest, I am not sure, so don't wanna make any false statements 😬

**Comment by milosgajdos** [comment]
Ok, I've cleaned up my last night's horrible code and I think we can proceed to getting this reviewed and iterated on.

PTAL @thaJeztah @dmcgowan @corhere @Jamstah

File: docs/spec/auth/jwt.md
Comment (Jamstah) [reply docs/spec/auth/jwt.md:71]: Does go-jose handle the kid at all, or does it leave it up to the consumer?

Comment (milosgajdos) [reply docs/spec/auth/jwt.md:71]: It does it for the `JSONWebKey` indeed https://pkg.go.dev/github.com/go-jose/go-jose/v3#JSONWebKey, but note that we also need to generate it from the provided trusted keys (see the constructor of the `accessController` and the token verification checks, etc.)

Comment (Jamstah) [reply docs/spec/auth/jwt.md:71]: I was really asking, if we didn't use the libtrust format, would JOSE provide anything useful.

Personally, I think removing the libtrust format makes sense.

File: registry/auth/token/crypto.go
Comment (Jamstah) [inline registry/auth/token/crypto.go:89]: Do we have a test for this? Was there something in libtrust that tested this?

**Review by Jamstah — CHANGES_REQUESTED** [review]
I still think we need to separate the test code more cleanly.

I'd like to remove the libtrust kid format if we can. I know we're no longer depending on the package, but consumers would either need to depend on the libtrust package to make the kid, or also have their own code. It doesn't seem to be a standards thing.

File: registry/auth/token/crypto.go
Comment (milosgajdos) [reply registry/auth/token/crypto.go:21]: I was suggesting this in one of the earlier comments 😄

File: registry/auth/token/crypto.go
Comment (milosgajdos) [reply registry/auth/token/crypto.go:89]: We don't I'm happy to add one, indeed.

File: registry/auth/token/crypto.go
Comment (milosgajdos) [reply registry/auth/token/crypto.go:21]: See  here 😄 https://github.com/distribution/distribution/pull/4096#discussion_r1350145871

Comment (Jamstah) [reply registry/auth/token/crypto.go:21]: That was certificates.go, this is crypto.go. crypto has a couple of methods which are used in runtime code, and everything else is only used in test, from what I can see.

Comment (milosgajdos) [reply registry/auth/token/crypto.go:21]: Yeah, my comment wasn't clear but I meant it more generally - sorry, should've been clearer.

Comment (Jamstah) [reply registry/auth/token/crypto.go:21]: No worries, we're on the same page :)

**Comment by Jamstah** [comment]
FYI, the export regulation disclaimer worried me too, so I raised this: https://github.com/go-jose/go-jose/issues/61

File: registry/auth/token/crypto.go
Comment (milosgajdos) [reply registry/auth/token/crypto.go:89]: Actually, it seems like there was no test for this in the `libtrust` repo https://github.com/search?q=repo%3Adocker%2Flibtrust%20keyIDFromCryptoKey&type=code.

File: registry/auth/token/crypto.go
Comment (milosgajdos) [reply registry/auth/token/crypto.go:33]: This is actually not easy to rename - there is the `keyIDFromCryptoKey` func we call in `accesscontroller` to get the public key fingerprint and store it in the trusted keys.

**Comment by milosgajdos** [comment]
> I'd like to remove the libtrust kid format if we can. I know we're no longer depending on the package, but consumers would either need to depend on the libtrust package to make the kid, or also have their own code. It doesn't seem to be a standards thing.

What are your thoughts @dmcgowan? I kinda agree but we need some replacement for that if we do rip that out.

File: registry/auth/token/certificates.go
Comment (milosgajdos) [reply registry/auth/token/certificates.go]: I've moved the contents of this file to `token_test.go`. The code is not used anywhere else so let's avoid keeping that code in a dedicated file.

File: registry/auth/token/crypto.go
Comment (milosgajdos) [reply registry/auth/token/crypto.go:28]: https://github.com/distribution/distribution/pull/4096#discussion_r1352006729

**Comment by Jamstah** [comment]
> > I'd like to remove the libtrust kid format if we can. I know we're no longer depending on the package, but consumers would either need to depend on the libtrust package to make the kid, or also have their own code. It doesn't seem to be a standards thing.
> 
> What are your thoughts @dmcgowan? I kinda agree but we need some replacement for that if we do rip that out.

Do we need a replacement?

The kid is really for picking a key from a JWKS that is provided by an issuer. I don't think that inventing one from the key material itself is valuable unless someone is coding an issuer implementation specifically for us that also matches (like what libtrust did and expected).

I guess one option would be to ask the user to specify the kid for each public cert. Another option would be to accept a JWKS url and pick up the kid/cert pairs from that. A final slightly less secure option would be to ignore the kid and test all the roots.

File: registry/auth/token/crypto.go
Comment (Jamstah) [reply registry/auth/token/crypto.go:89]: Shame, but I guess it won't matter if I convince you to remove the libtrust kid format :)

File: registry/auth/token/crypto.go
Comment (Jamstah) [reply registry/auth/token/crypto.go:33]: vscode tells me the only reference is `TestNewAccessControllerPemBlock` in `token_test.go`.

`keyIDFromCryptoKey` uses `x509.MarshalPKIXPublicKey` directly.

File: registry/auth/token/crypto.go
Comment (Jamstah) [reply registry/auth/token/crypto.go:28]: According to vscode: called from functions in certificate.go (which are all test), PEMBlock (see https://github.com/distribution/distribution/pull/4096#discussion_r1352006729), and token_test.go.

**Comment by milosgajdos** [comment]
> Do we need a replacement?

Well `accesscontroller` maintains a map of trusted publick keys https://github.com/distribution/distribution/blob/f7b3869062dbbfdf0260d223bac90a69418c3028/registry/auth/token/accesscontroller.go#L135

Those are optionally supplied by config https://github.com/distribution/distribution/blob/f7b3869062dbbfdf0260d223bac90a69418c3028/registry/auth/token/accesscontroller.go#L184

**Comment by Jamstah** [comment]
The key of the map is what we're talking about though, its the libtrust format. Its only valuable if the token issuer uses the same mapping for the keys in its JWKS. Considering that no one should be using libtrust any more, that's a big ask.

We really need the user to specify kid/root pairs so they can match them to their issuer. Even better would be to accept a JWKS and use that instead.

**Comment by milosgajdos** [comment]
> The key of the map is what we're talking about though

Yes very much so - too many overloaded terms!

**Comment by Jamstah** [comment]
I think this might be the right option:
- If the user specifies roots, then we only verify tokens that are x5c.
- If the user specifies a file that contains a JWKS, then we give it to go-jose and let it do the work.

**Comment by milosgajdos** [comment]
> I think this might be the right option:
> 
>     * If the user specifies roots, then we only verify tokens that are x5c.
> 
>     * If the user specifies a file that contains a JWKS, then we give it to go-jose and let it do the work.
> 
> 
> So, `auth.rootCertBundle` and `auth.jwks` in the configuration.

Correct me if I'm wrong but there should be an option to specify both, right 🤔 It should be your choice what tokens your registry trusts. I may be overthinking this though...

**Comment by Jamstah** [comment]
Yes, I don't see why not. If there are both roots and a JWKS configured, use them both.

Or, if there is an x5c in the token, we could verify that with any roots found in the JWKS (which is like our behaviour now). Then we would only have the JWKS option in the config.

File: registry/auth/token/crypto.go
Comment (milosgajdos) [reply registry/auth/token/crypto.go]: Fixed: https://github.com/distribution/distribution/pull/4096/commits/88e43a4217e81f36b2f0266bf05033071308bebd

**Comment by milosgajdos** [comment]
Alright, so I got a first step in the direction we agreed on - unless I misunderstood 🙃 

PTAL @Jamstah Mind you, it's a bit of a ghetto code that will need cleaning up, but would like your take on this 😬 

Also, if @corhere has some spare cycles, that'd be dope  (I'm still expecting him to open that `context` cleanup PR he keeps promising me 😏)

(yes we will still need to update docs with the `jwks` option, but I postponed it until we've agreed on the code 😄 )

File: registry/auth/token/crypto.go
Comment (corhere) [reply registry/auth/token/crypto.go]: [Not quite fixed yet.](https://github.com/docker/libtrust/blob/aabc10ec26b754e797f9028f4589c5b7bd90dc20/ec_key.go#L341) EC private keys should remain type `"EC PRIVATE KEY"`.

Comment (milosgajdos) [reply registry/auth/token/crypto.go]: Ah! Gotchya. I misunderstood. Ok, will fix in a followup.

**Comment by corhere** [comment]
> (I'm still expecting him to open that `context` cleanup PR he keeps promising me 😏)

Hey now, [I only promised to _review_ the context cleanup PR you were going to open](https://github.com/distribution/distribution/pull/4066#discussion_r1337467908) @milosgajdos

File: registry/auth/token/crypto.go
Comment (milosgajdos) [reply registry/auth/token/crypto.go]: Also, worth mentioning, @corhere this is only used in tests, but I'll updarte

**Comment by milosgajdos** [comment]
> Hey now, [I only promised to _review_ the context cleanup PR you were going to open](https://github.com/distribution/distribution/pull/4066#discussion_r1337467908) @milosgajdos

Saw right through my innocuous gaslighting 🙃

File: registry/auth/token/crypto.go
Comment (milosgajdos) [reply registry/auth/token/crypto.go]: libtrust: https://github.com/docker/libtrust/blob/aabc10ec26b754e797f9028f4589c5b7bd90dc20/key.go#L146-L152
Last commit: https://github.com/distribution/distribution/pull/4096/commits/8f6df3fcee1b2a18b1fad1aaf3769d7ea44f2764

**Comment by Jamstah** [comment]
Should we be looking at something like this?

https://pkg.go.dev/github.com/s12v/go-jwks

**Comment by milosgajdos** [comment]
> Should we be looking at something like this?
> 
> https://pkg.go.dev/github.com/s12v/go-jwks

You mean for rotation? Yeah, seems reasonable. Let's rip `libtrust` out first though 😄

**Review by Jamstah — COMMENTED** [review]
We can rip out a load of the libtrust crypto from the test files too.

We don't have to, but it reduces the complexity.

File: registry/auth/token/accesscontroller.go
Comment (milosgajdos) [reply registry/auth/token/accesscontroller.go:266]: But thats the correct behaviour though, right? If you don't provide trusted keys, what do you verify against? 🤔

**Comment by milosgajdos** [comment]
> We can rip out a load of the libtrust crypto from the test files too.
> 
> We don't have to, but it reduces the complexity.
> 
> [Jamstah@f485172](https://github.com/Jamstah/distribution/commit/f485172f6d7cf9371fb0529b6cd12b246804f794)

I was going to look into ripping that `crypto_test.go` file out in the followup. There is more cleanup to be done beyond that, I feel 😅 Nice! I'll [shamelessly] adopt these changes 😄

File: registry/auth/token/accesscontroller.go
Comment (Jamstah) [reply registry/auth/token/accesscontroller.go:266]: By "this if will fail" I mean the conditional statement will be false and no error will be raised. Unless I read it wrong.

Comment (milosgajdos) [reply registry/auth/token/accesscontroller.go:266]: Ah, I see what you mean. Yeah, nice spot! Needs fixing.

Comment (milosgajdos) [reply registry/auth/token/accesscontroller.go]: fix: https://github.com/distribution/distribution/pull/4096/commits/bbba63f4455ddbdf40a161f653df492722c4dab8

**Comment by thaJeztah** [comment]
I'll try to find some time to review, but with a big disclaimer that this area is not really my expertise. I did ping @jcarter3 and @technicallyjosh on our Slack, as they are more familiar with this, and they were definitely in favor of removing libtrust (but don't know to what extent they reviewed the change in-depth).

**Comment by milosgajdos** [comment]
@thaJeztah we need @dmcgowan here primarly. He was the OG creator so he likely understands ins and outs of this code.

**Comment by dmcgowan** [comment]
Agreed that removing the KID definition here makes sense to match the spec, since it is "unspecified".  It should be up to the token service and jwk creator to ensure it matches using whatever format or criteria they determine to be the most secure. 

The key sets from an endpoint seem interesting as well as a future change and align nicely with the distribution architecture of horizontally scaled registry instances using a token service which owns the token private key and could potentially serve that as an endpoint. This would make key rotation much simpler. Going along with that, supporting multiple signing keys could aid in that rotation.

**Review by jcarter3 — APPROVED** [review]

**Comment by milosgajdos** [comment]
@Jamstah anything we're missing here that's preventing you from ✅ ? Other than the [inevitable] squash 😄

**Comment by thaJeztah** [comment]
Squash (or cleanup of some of the commits at least) would be nice 😅

**Comment by milosgajdos** [comment]
> Squash (or cleanup of some of the commits at least) would be nice 😅

Yes, just wanna make sure we're done updating before I squash and merge. Otherwise all the discussion in this terrible GH Review UI will be lost in the darkness of our memories

**Review by Jamstah — APPROVED** [review]

**Comment by Jamstah** [comment]
Sorry, had forgotten I had a request changes open!

Lgtm.

**Pull request merged**
